### PR TITLE
Offence matching

### PR DESCRIPTION
**Pull request by richardcane**
Rich: I don't want to let this PR grow any more. There are some `TODO`s in this. It would be better to iterate on these.

### CODE REVIEW

File: cypress/component/OffenceMatcher/SingleCourtCase.cy.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        </CourtCaseContext.Provider>
+      )
+    })
+
+    it("has a default placeholder", () => {
+      cy.get("select").should("have.value", null)
+      cy.get("select").find(":selected").should("have.text", "Select an offence")
+    })
+
+    it("displays CCR as an <optgroup>", () => {
```
Comment (chubberlisk) [inline cypress/component/OffenceMatcher/SingleCourtCase.cy.tsx:25]:
```suggestion
    it("groups dropdown options by CCR", () => {
```

File: cypress/component/OffenceMatcher/SingleCourtCase.cy.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      cy.get("select").children("optgroup").eq(0).should("have.attr", "label", "97/1626/008395Q")
+      cy.get("select").children("optgroup").eq(0).contains("option", "TH68006")
+    })
+
+    it("displays matchable offences with an <option>", () => {
```
Comment (chubberlisk) [inline cypress/component/OffenceMatcher/SingleCourtCase.cy.tsx:30]:
```suggestion
    it("displays matchable offences as dropdown options", () => {
```

File: cypress/component/OffenceMatcher/SingleCourtCase.cy.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    it("displays offences in the correct format", () => {
+      cy.get("option")
+        .contains("TH68006")
+        .invoke("text")
+        .then((text) => {
+          expect(text).equals("001 - TH68006")
+        })
+    })
+
+    it("displays an <option> for 'Added in court' as the last option", () => {
```
Comment (chubberlisk) [inline cypress/component/OffenceMatcher/SingleCourtCase.cy.tsx:45]:
```suggestion
    it("displays 'Added in court' as the last option", () => {
```

**Review by chubberlisk — APPROVED** [review]
:shipit: 

I've just lightly reviewed this PR with the idea that it's feature flag and we'll be iterating on it very soon!

**Review by chubberlisk — DISMISSED** [review]

File: src/components/OffenceMatcher.tsx [hunk 4]
```diff
@@ -1,39 +1,59 @@
...
   }
 
-  const isAlreadySelected = (sequenceNumber: string) =>
-    !!amendments.offenceReasonSequence?.find((x) => x.value === sequenceNumber && x.offenceIndex !== offenceIndex)
+  const offenceAlreadySelected = (sequenceNumber: number) =>
+    !!amendments.offenceReasonSequence?.find((a) => a.value === sequenceNumber && a.offenceIndex !== offenceIndex)
 
-  // TODO: load manually selected value if exists (just load updated aho always?)
-  // TODO: prevent matching twice
   // TODO: match dates
```
Comment (richrace) [inline src/components/OffenceMatcher.tsx:51]: Shouldn't this be done in Core?

**Review by emadkaramad — APPROVED** [review]

**Pull request merged**